store/tikv/backoff.go

```diff
@@ -15,431 +15,54 @@ package tikv
-		b.fn = make(map[BackoffType]func(context.Context, int) int)
-	}
-	f, ok := b.fn[typ]
-	if !ok {
-		f = typ.createFn(b.vars)
-		b.fn[typ] = f
-	}
-
-	realSleep := f(b.ctx, maxSleepMs)
-	typ.metric().Observe(float64(realSleep) / 1000)
-	b.totalSleep += realSleep
-	if b.backoffSleepMS == nil {
-		b.backoffSleepMS = make(map[BackoffType]int)
-	}
-	b.backoffSleepMS[typ] += realSleep
-	if b.backoffTimes == nil {
-		b.backoffTimes = make(map[BackoffType]int)
-	}
-	b.backoffTimes[typ]++
-
-	stmtExec := b.ctx.Value(util.ExecDetailsKey)
-	if stmtExec != nil {
-		detail := stmtExec.(*util.ExecDetails)
-		atomic.AddInt64(&detail.BackoffDuration, int64(realSleep)*int64(time.Millisecond))
-		atomic.AddInt64(&detail.BackoffCount, 1)
-	}
-
-	if b.vars != nil && b.vars.Killed != nil {
-		if atomic.LoadUint32(b.vars.Killed) == 1 {
-			return tikverr.ErrQueryInterrupted
-		}
-	}
-
-	var startTs interface{}
-	if ts := b.ctx.Value(TxnStartKey); ts != nil {
-		startTs = ts
-	}
-	logutil.Logger(b.ctx).Debug("retry later",
-		zap.Error(err),
-		zap.Int("totalSleep", b.totalSleep),
-		zap.Int("maxSleep", b.maxSleep),
-		zap.Stringer("type", typ),
-		zap.Reflect("txnStartTS", startTs))
-	return nil
-}
-
-func (b *Backoffer) String() string {
-	if b.totalSleep == 0 {
-		return ""
-	}
-	return fmt.Sprintf(" backoff(%dms %v)", b.totalSleep, b.types)
-}
-
-// Clone creates a new Backoffer which keeps current Backoffer's sleep time and errors, and shares
-// current Backoffer's context.
-func (b *Backoffer) Clone() *Backoffer {
-	return &Backoffer{
-		ctx:        b.ctx,
-		maxSleep:   b.maxSleep,
-		totalSleep: b.totalSleep,
-		errors:     b.errors,
-		vars:       b.vars,
-	}
-}
-
-// Fork creates a new Backoffer which keeps current Backoffer's sleep time and errors, and holds
-// a child context of current Backoffer's context.
-func (b *Backoffer) Fork() (*Backoffer, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(b.ctx)
-	return &Backoffer{
-		ctx:        ctx,
-		maxSleep:   b.maxSleep,
-		totalSleep: b.totalSleep,
-		errors:     b.errors,
-		vars:       b.vars,
-	}, cancel
-}
-
-// GetVars returns the binded vars.
-func (b *Backoffer) GetVars() *kv.Variables {
-	return b.vars
-}
-
-// GetTotalSleep returns total sleep time.
-func (b *Backoffer) GetTotalSleep() int {
-	return b.totalSleep
-}
-
-// GetTypes returns type list.
-func (b *Backoffer) GetTypes() []fmt.Stringer {
-	return b.types
-}
-
-// GetCtx returns the binded context.
-func (b *Backoffer) GetCtx() context.Context {
-	return b.ctx
+// NewBackoffer creates a Backoffer with maximum sleep time(in ms).
+func NewBackoffer(ctx context.Context, maxSleep int) *Backoffer {
+	return retry.NewBackoffer(ctx, maxSleep)
 }
 
-// GetBackoffTimes returns a map contains backoff time count by type.
-func (b *Backoffer) GetBackoffTimes() map[BackoffType]int {
-	return b.backoffTimes
+// TxnStartKey is a key for transaction start_ts info in context.Context.
+func TxnStartKey() interface{} {
+	return retry.TxnStartKey
```

Comment:
maybe `var TxnStartKey = retry.TxnStartKey`.

Reply:
I do not think so, consider the following case:

```
retry.TxnStartKey = 1
var TxnStartKey = retry.TxnStartKey // TxnStartKey is 1
retry.TxnStartKey = 2 // retry.TxnStartKey 
fmt.Println(TxnStartKey) // here we wish TxnStartKey should be 2 while it is 1.

```